### ISSUE TITLE
upgrade to 1.2.4 broken the service

### ISSUE THREAD

**Issue by EugenMayer** [osixia/docker-openldap#307]
Using the image for a very long while now, today after upgrading to 1.2.4 ( from 1.2.2 ) it fails with

```

*** Killing all processes...
*** CONTAINER_LOG_LEVEL = 3 (info)
*** Search service in CONTAINER_SERVICE_DIR = /container/service :
*** link /container/service/:ssl-tools/startup.sh to /container/run/startup/:ssl-tools
*** failed to link /container/service/:ssl-tools/startup.sh to /container/run/startup/:ssl-tools: [Errno 17] File exists
*** link /container/service/slapd/startup.sh to /container/run/startup/slapd
*** failed to link /container/service/slapd/startup.sh to /container/run/startup/slapd: [Errno 17] File exists
*** link /container/service/slapd/process.sh to /container/run/process/slapd/run
*** directory /container/run/process/slapd already exists
*** failed to link /container/service/slapd/process.sh to /container/run/process/slapd/run : [Errno 17] File exists
*** Set environment for startup files
*** Environment files will be proccessed in this order :
Caution: previously defined variables will not be overriden.
/container/environment/99-default/default.startup.yaml
/container/environment/99-default/default.yaml

To see how this files are processed and environment variables values,
run this container with '--loglevel debug'
*** Running /container/run/startup/:ssl-tools...
*** Running /container/run/startup/slapd...



Warning: LDAP_BACKEND environment variable is set to mdb but hdb backend is detected.
Going to use hdb as LDAP_BACKEND. Set LDAP_BACKEND=hdb to discard this message.
https://github.com/osixia/docker-openldap#set-your-own-environment-variables



Start OpenLDAP...
Waiting for OpenLDAP to start...
Add TLS config...
Disable replication config...
*** /container/run/startup/slapd failed with status 49
```

Any hints on what is causing this?

I am running this image in my rancher catalog, the docker-compose file looks like this https://github.com/EugenMayer/docker-rancher-extra-catalogs/blob/master/templates/openldap/19/docker-compose.yml#L3

are there any changes to the configuration or such?

Those are the volumes 
https://github.com/EugenMayer/docker-rancher-extra-catalogs/blob/master/templates/openldap/19/docker-compose.yml#L31

**Comment by EugenMayer** [comment]
My env-vars ( partially redacted ) looklike 

```
"LDAP_ADMIN_PASSWORD": "XXXXXXXXXXXX",
"LDAP_CONFIG_PASSWORD": "XXXXXXXXXXXX",
"LDAP_DOMAIN": "auth.YYYYYYYYYYYYYY",
"LDAP_ORGANISATION": "ZZZZZZZZZ",
"LDAP_READONLY_USER": "true",
"LDAP_READONLY_USER_PASSWORD": "ZZZZZZZ",
"LDAP_READONLY_USER_USERNAME": "readonly",
"LDAP_REPLICATION": "false",
"LDAP_SSL_HELPER_PREFIX": "ldap",
"LDAP_TLS": "true",
"LDAP_TLS_CA_CRT_FILENAME": "ca.pem",
"LDAP_TLS_CIPHER_SUITE": "NORMAL:SECURE256:-VERS-SSL3.0",
"LDAP_TLS_CRT_FILENAME": "cert.pem",
"LDAP_TLS_ENFORCE": "false",
"LDAP_TLS_KEY_FILENAME": "privkey.pem",
"LDAP_TLS_PROTOCOL_MIN": "3.1",
"LDAP_TLS_VERIFY_CLIENT": "never"
```

**Comment by EugenMayer** [comment]
It is odd that the server brings this `mdb` issue while the ENV var selects `hdb` - does the configuration gets "recreated" on every container-creation or just on the first container init? I mean, changing those ENV vars has any effect at all after bootstrapping that service one ( and having persistent volumes )?

**Comment by EugenMayer** [comment]
Ok activating the trace mode helped ..

```
5ca76231 conn=1004 fd=17 ACCEPT from IP=127.0.0.1:34594 (IP=127.0.0.1:389)
5ca76231 conn=1004 op=0 BIND dn="cn=admin,dc=auth,dc=kw,XXXXXX" method=128
5ca76231 conn=1004 op=0 RESULT tag=97 err=49 text=
5ca76231 conn=1004 op=1 UNBIND
5ca76231 conn=1004 fd=17 closed
ldap_bind: Invalid credentials (49)
*** /container/run/startup/slapd failed with status 49
```

But why is that happening suddenly ;/

**Comment by EugenMayer** [comment]
Ok i was able to sort this out, but iam not sure this is a regression of https://github.com/osixia/docker-openldap/commit/b25e0a50d33507557a7a83164a19b080fdd03255#diff-04c6e90faac2675aa89e2176d2eec7d8

The point is, that i had set
 `LDAP_DOMAIN=auth.kw.mycompany.com` 
and i was not setting `LDAP_BASE_DN` at all.

Still, my dn of my admin user was `cn=admin,dc=mycompany,dc=com`. When uprading to `1.2.4` openldap tried to connect to the database using the admin dn

`cn=admin,dc=auth,dc=kw,dc=mycompany,dc=com` ... so now with `dc=auth,dc=kw` in the middle .. and this failed with invalid credentials ( since this user does not exist ) ..

Setting `LDAP_BASE_DN` to `dc=mycompany,dc=com` fixed this, since now `cn=admin,dc=mycompany,dc=com` was used

---

Summing this up, keeping this issue open so we can understand
a) is this just a configuration issue by me ( close the issue then )
b) is this a regression / bug .. why did it happen with 1.2.4 suddenly - what has changed
c) do we need to inform other users / in the readme?

Thanks!

**Comment by EugenMayer** [comment]
Any way i could move this ticket forward / provide better informations?

**Comment by vityafx** [comment]
I have exactly the same problem, how did you enable this trace mode?

**Comment by EugenMayer** [comment]
https://github.com/EugenMayer/docker-rancher-extra-catalogs/blob/master/templates/openldap/19/docker-compose.yml#L3

"--loglevel ${SLAPD_LOG_LEVEL}"

**Comment by vityafx** [comment]
Thanks. My problem was with TLS certificates.

**Comment by obourdon** [comment]
@EugenMayer do you mean that before 1.2.4, setting LDAP_DOMAIN=auth.kw.mycompany.com
did not have any impact on the admin DN which could still be set to cn=admin,dc=mycompany,dc=com automatically ??? Or did you add this LDAP_DOMAIN later on. Please clarify

**Comment by EugenMayer** [comment]
@obourdon sure

Facts:
- prior 1.2.2 i never set `LDAP_BASE_DN` at all.
- i always had `LDAP_DOMAIN` set to `auth.kw.mycompany.com`
- starting with 1.2.4 ( i did not test 1.2.3 though, skipped it ) i have to set `LDAP_BASE_DN=mycompany.com` to be able to let the boostrap process understand to bind with `cn=admin,dc=mycompany,dc=com`.

Not setting `LDAP_BASE_DN` (in 1.2.4)  will rather use `cn=admin,dc=auth,dc=kw,dc=mycompany,dc=com` as bind user and that will fail, since my user is `cn=admin,dc=mycompany,dc=com`

Did that help?

**Comment by obourdon** [comment]
@EugenMayer yes it definitely helps thanks. Will dig into this further and let you know about findings

**Comment by Y0hy0h** [comment]
It appears that I'm encountering a similar issue. When the `LDAP_DOMAIN` (let's assume its `example`, without any TLD) does not match the `LDAP_BASE_DN` (let's assume its `dc=example,dc=org`, with TLD), the wrong one seems to be used for creating the base.

```
ldap        | ldap_add: Server is unwilling to perform (53)
ldap        |   additional info: no global superior knowledge
ldap        | adding new entry "ou=Group,dc=example,dc=org"
ldap        | 5d17ef6d conn=1027 fd=18 ACCEPT from IP=127.0.0.1:42162 (IP=127.0.0.1:389)
ldap        | 5d17ef6d conn=1027 op=0 BIND dn="cn=admin,dc=example,dc=org" method=128
ldap        | 5d17ef6d conn=1027 op=0 RESULT tag=97 err=49 text=
ldap        | 5d17ef6d conn=1027 op=1 UNBIND
ldap        | 5d17ef6d conn=1027 fd=18 closed
ldap        | ldap_bind: Invalid credentials (49)
ldap        | *** /container/run/startup/slapd failed with status 49
ldap        |
ldap        | *** Run commands before finish...
ldap        | *** Killing all processes...
ldap        | 5d17ef6d daemon: shutdown requested and initiated.
ldap        | 5d17ef6d slapd shutdown: waiting for 0 operations/tasks to finish
ldap        | 5d17ef6d slapd stopped.
```

Especially `no global superior knowledge` appears to me to hint at some bug in the start up script, that doesn't correctly handle differing `LDAP_DOMAIN` and `LDAP_BASE_DN`.

**Comment by Y0hy0h** [comment]
I managed to pinpoint it to [this line](https://github.com/osixia/docker-openldap/blob/stable/image/service/slapd/startup.sh#L105). If I manually enter a domain that is consistent with the `LDAP_BASE_DN`, it works. E. g. I set it to `example.org` when my `LDAP_BASE_DN` is `dc=example,dc=org`. Apparently, the setup of the admin is left to `slapd`, and it uses the domain to derive the base dn, which causes problems when the domain is different from the desired base dn.

Unfortunately I couldn't find any documentation on whether `slapd` allows to configure the base dn the same way as this image does with e. g. a `slapd/base_dn` setting. I also couldn't find documentation on how precisely `slapd` handles the `slapd/domain` configuration, so I'm not sure what happens if we change it.

Does anyone know how to handle this?
